### DISCUSSION TITLE
set new window urgent

### DISCUSSION THREAD

**Discussion by jekanev** [qtile/qtile#3154] [Q&A] [answered]
I want all new windows to be urgent if it is started outside the current group
setting
focus_on_window_activation = 'urgent'
dont give expected behavior in all windows
i write hook

```
@libqtile.hook.subscribe.client_managed
def func(c):
   if qtile.current_group != c.group:
       subprocess.run(["~/.config/urgent.sh"])
```

```
urgent.sh
#!/bin/bash
wmid=$(wmctrl -l | tail -n1 | cut -d" " -f1)
wmctrl -i -r "$wmid" -b add,demands_attention
```

it seems to work as i need, wanted to ask if it is possible to make the window urgent without using wmctrl and bash script.

**Comment by elParaguayo** [comment, marked as answer]
What happens if you just do:
``` python
@libqtile.hook.subscribe.client_managed
def func(c):
   if qtile.current_group != c.group:
       c.urgent = True
```

This may not trigger the events that would normally happen if a window identifies itself as urgent.

**Reply by jekanev** [reply]
i set c.urgent = True, but this did not lead to the group icon changing color to red and the lazy.next_urgent()  function does not switch to this window. well seems wmctrl is only one way.

**Reply by elParaguayo** [reply]
Hang on. I'm not done yet. There may be other ways in Qtile. Let me look some more.

**Reply by jekanev** [reply]
got it, thanks any way.

**Reply by elParaguayo** [reply]
You can probably fix the groupbox by adding another line `hook.fire("client_urgent_hint_changed", c)`. I'll look at the `next_urgent` code next,

**Reply by elParaguayo** [reply]
Actually, just testing this. It's not actually setting `urgent` at all (when I check the value, it's still `False`). More looking needed.

**Reply by elParaguayo** [reply]
It's weird. Setting `urgent = True` should work but that flag is getting reset somewhere and I can't see why.

**Reply by jekanev** [reply]
git it, well bash script seems work ok for me, initially I thought that replacing the bash script would be easier. thanks for help any way, personally for me this question answered atm, should i make this discussion new issue or i can just mark answered?

**Reply by elParaguayo** [reply]
You can mark as answered if you're happy.

I may keep looking into this as I'm not sure why it doesn't work!

**Reply by jekanev** [reply]
ok ty.

**Comment by Koljasha** [comment]
@elParaguayo I have the same question.
I would like to Qtile worked like i3 in "smart" mode: https://i3wm.org/docs/userguide.html#focus_on_window_activation
I changed the variable 'focus_on_window_activation', but Urgent Flag does not work correctly.
Now I use a hook on the script above and the variable in the 'urgent'. In my opinion it works correctly. But it is not correct, it should only work through with Qtile.

**Reply by elParaguayo** [reply]
I don't think it will work if you're trying to set "urgent" in the hook as the window's already there. Setting urgent true won't cause the window to be focused. Why not just focus the window from your hook instead?

**Comment by Koljasha** [comment]
@elParaguayo Sorry. Probably I mistakenly translated from translate.google.com )
My task: I while in the group "1" launch an application that opens in the group "5" -> Group "5" became "Urgent".
So works in i3wm with "smart" (default).
focus_on_window_activation for some reason does not work.
c.urgent = True in hook does not work.
c.hints["urgent"] = True
At the moment, it is possible to obtain the expected option only as described above (wmctrl+hook).
Thanks @jekanev for this discussion and script.

**Reply by elParaguayo** [reply]
Correct. Neither of these options:
``` python
c.urgent = True in hook does not work.
c.hints["urgent"] = True
```
will have any impact on the `focus_on_window_activation`.

**Comment by elParaguayo** [comment]
Got it!
``` python
@hook.subscribe.client_managed
def make_urgent(window):
    atom = set([qtile.core.conn.atoms["_NET_WM_STATE_DEMANDS_ATTENTION"]])
    prev_state = set(window.window.get_property("_NET_WM_STATE", "ATOM", unpack=int))
    new_state = prev_state | atom
    window.window.set_property("_NET_WM_STATE", list(new_state))
```
You can adjust this to only change windows that meet certain conditions. As it is, every window will be made urgent.

**Reply by Koljasha** [reply]
Yes, It works for me!
Thank you very much! Happy New Year!
